nowcasting_utils/models/loss.py

```diff
@@ -43,3 +51,310 @@ def get_mse_exp(self, output, target):
     def get_mae_exp(self, output, target):
         """Loss function weighted MAE"""
         return torch.sum(self.weights * torch.abs(output - target))
+
+
+class SSIMLoss(nn.Module):
+    def __init__(self, convert_range: bool = False, **kwargs):
+        """
+        SSIM Loss, optionally converting input range from [-1,1] to [0,1]
+        Args:
+            convert_range: Convert input from -1,1 to 0,1 range
+            **kwargs: Kwargs to pass through to SSIM
+        """
+        super(SSIMLoss, self).__init__()
+        self.convert_range = convert_range
+        self.ssim_module = SSIM(**kwargs)
+
+    def forward(self, x: torch.Tensor, y: torch.Tensor):
+        if self.convert_range:
+            x = torch.div(torch.add(x, 1), 2)
+            y = torch.div(torch.add(y, 1), 2)
+        return 1.0 - self.ssim_module(x, y)
+
+
+class MS_SSIMLoss(nn.Module):
+    def __init__(self, convert_range: bool = False, **kwargs):
+        """
+        Multi-Scale SSIM Loss, optionally converting input range from [-1,1] to [0,1]
+        Args:
+            convert_range: Convert input from -1,1 to 0,1 range
+            **kwargs: Kwargs to pass through to MS_SSIM
+        """
+        super(MS_SSIMLoss, self).__init__()
+        self.convert_range = convert_range
+        self.ssim_module = MS_SSIM(**kwargs)
+
+    def forward(self, x: torch.Tensor, y: torch.Tensor):
+        if self.convert_range:
+            x = torch.div(torch.add(x, 1), 2)
+            y = torch.div(torch.add(y, 1), 2)
+        return 1.0 - self.ssim_module(x, y)
+
+
+class SSIMLossDynamic(nn.Module):
+    def __init__(self, convert_range: bool = False, **kwargs):
+        """
+        SSIM Loss on only dynamic part of the images, optionally converting input range from [-1,1] to [0,1]
+
+        In Mathieu et al. to stop SSIM regressing towards the mean and predicting only the background, they only
+        run SSIM on the dynamic parts of the image. We can accomplish that by subtracting the current image from the future ones
+
+        Args:
+            convert_range: Whether to convert from -1,1 to 0,1 as required for SSIM
+            **kwargs: Kwargs for the ssim_module
+        """
+        super(SSIMLossDynamic, self).__init__()
+        self.convert_range = convert_range
+        self.ssim_module = MS_SSIM(**kwargs)
+
+    def forward(self, curr_image: torch.Tensor, x: torch.Tensor, y: torch.Tensor):
+        if self.convert_range:
```

Comment:
worth adding function descirption, for example what curr_image is

maybe worth using 'current_image' - fully verbose variables